server/gfmd/user.c

```diff
@@ -1590,4 +1782,252 @@ gfm_server_user_info_remove(struct peer *peer, int from_client, int skip)
+	}
+	giant_unlock();
+	free(auth_user_id_type_str);
+	free(auth_user_id);
+	return (e);
+}
+
+gfarm_error_t
+gfm_server_user_auth_get(
+	struct peer *peer, int from_client, int skip)
+{
+	gfarm_error_t e;
+	char *username;
+	char *auth_user_id_type_str;
+	enum auth_user_id_type auth_user_id_type;
+	struct user *u;
+	struct process *process;
+	int name_with_tenant;
+	static const char diag[] = "GFM_PROTO_USER_AUTH_GET";
+
+	e = gfm_server_get_request(peer, diag, "ss",
+				   &username, &auth_user_id_type_str);
+	if (e != GFARM_ERR_NO_ERROR) {
+		gflog_debug(GFARM_MSG_UNFIXED,
+		    "%s request: %s", diag, gfarm_error_string(e));
+		return (e);
+	}
+	if (skip) {
+		free(username);
+		free(auth_user_id_type_str);
+		return (GFARM_ERR_NO_ERROR);
+	}
+
+	e = gfarm_auth_user_id_type_parse(auth_user_id_type_str,
+					  &auth_user_id_type);
+	if (e != GFARM_ERR_NO_ERROR) {
+		gflog_debug(GFARM_MSG_UNFIXED, "invalid auth_user_id_type");
+		return (e);
+	}
+
+	/* XXX FIXME too long giant lock */
+	giant_lock();
+
+	if ((e = rpc_name_with_tenant(peer, from_client,
+	    &name_with_tenant, &process, diag)) != GFARM_ERR_NO_ERROR) {
+		/* nothing to do */
+	} else if ((u = user_tenant_lookup(username)) == NULL) {
+		e = GFARM_ERR_NO_SUCH_USER;
+	} else
+		e = GFARM_ERR_NO_ERROR;
+
+	if (e != GFARM_ERR_NO_ERROR) {
+		e = gfm_server_put_reply(peer, diag, e, "");
+	} else {
+		/* peer_get_user(peer) is not NULL if from_client */
+		e = gfm_server_put_reply(peer, diag, e, "s",
+			u->auth_user_id[auth_user_id_type]);
+	}
+	giant_unlock();
+	free(username);
+	free(auth_user_id_type_str);
+	return (e);
+}
+
+static gfarm_error_t
+gfm_server_user_auth_set_and_modify(struct peer *peer,
+				    int from_client, int skip, const char *diag)
+{
+	gfarm_error_t e;
+	char *username;
+	char *auth_user_id_type_str;
+	char *auth_user_id;
+	enum auth_user_id_type auth_user_id_type;
+	struct user *u;
+	struct process *process;
+	int name_with_tenant;
+
+	e = gfm_server_get_request(peer, diag, "sss",
+		   &username, &auth_user_id_type_str, &auth_user_id);
+	if (e != GFARM_ERR_NO_ERROR) {
+		gflog_debug(GFARM_MSG_UNFIXED,
+		    "%s request: %s", diag, gfarm_error_string(e));
+		return (e);
+	}
+	if (skip) {
+		free(username);
+		free(auth_user_id_type_str);
+		free(auth_user_id);
+		return (GFARM_ERR_NO_ERROR);
+	}
+
+	e = gfarm_auth_user_id_type_parse(auth_user_id_type_str,
+					  &auth_user_id_type);
+	if (e != GFARM_ERR_NO_ERROR) {
+		gflog_debug(GFARM_MSG_UNFIXED, "invalid auth_user_id_type");
+		return (e);
+	}
+
+	/* XXX FIXME too long giant lock */
+	giant_lock();
+
+	if ((e = rpc_name_with_tenant(peer, from_client,
+	    &name_with_tenant, &process, diag)) != GFARM_ERR_NO_ERROR) {
+		/* nothing to do */
+	} else if ((u = user_tenant_lookup(username)) == NULL) {
+		e = GFARM_ERR_NO_SUCH_USER;
+	} else
+		e = GFARM_ERR_NO_ERROR;
+
+	if (e == GFARM_ERR_NO_ERROR) {
+		free(u->auth_user_id[auth_user_id_type]);
+		u->auth_user_id[auth_user_id_type] = auth_user_id;
```

Comment:
This is not enough.
The followings are necessary:
* update the user_id_map hash table.
* the journal and the backend database should be updated by either db_user_auth_add() or db_user_auth_modify()